### DISCUSSION TITLE
Adding Composite Column Types to base unmapped class fails migration script generation

### DISCUSSION THREAD

**Discussion by anil-enable** [sqlalchemy/sqlalchemy#10379] [Usage Questions] [answered]
**Describe the bug**
I was going through the tutorial for adding composite column types where sets of columns can be associated with a single user-defined datatype (FYR https://docs.sqlalchemy.org/en/20/orm/composites.html#composite-column-types).

I followed the example and added the classes. Post this I ran the `alembic revision --autogenerate` to auto generate the migration scripts. This generated the migration scripts.

I tweaked the example a bit and moved this composite column definition to a base unmapped class and had a concrete mapped class extend this class. Post this the auto-generation fails.

**Expected behavior**
I expect the auto generation to work when I define a composite column on a base unmapped class. 

**To Reproduce**

Example 1 works, example 2 doesn't work.

```py
import dataclasses
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, composite, mapped_column

class Base(MappedAsDataclass, DeclarativeBase):
    pass

@dataclasses.dataclass
class Point:
    x: int
    y: int

# Example 1 
# as per tutorial in https://docs.sqlalchemy.org/en/20/orm/composites.html#composite-column-types
# added composite columns start and end
# auto-generation works for this

class Vertex(Base):
    __tablename__ = "vertices"

    id: Mapped[int] = mapped_column(primary_key=True)

    start: Mapped[Point] = composite(mapped_column("x1"), mapped_column("y1"))
    end: Mapped[Point] = composite(mapped_column("x2"), mapped_column("y2"))

# Example 2
# moved composite columns start, end to base unmapped class
# added a concrete mapped class that extends this class
# auto-generation fails for this

class BaseVertex(Base):
    __abstract__ = True

    id: Mapped[int] = mapped_column(primary_key=True)

    start: Mapped[Point] = composite(mapped_column("x1"), mapped_column("y1"))
    end: Mapped[Point] = composite(mapped_column("x2"), mapped_column("y2"))


class NamedVertex(BaseVertex):
    __tablename__ = "named_vertex"
    name: Mapped[str] = mapped_column()

```

**Error**

```
Traceback (most recent call last):
  File "/example/.venv/bin/alembic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/example/.venv/lib/python3.11/site-packages/alembic/config.py", line 630, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/example/.venv/lib/python3.11/site-packages/alembic/config.py", line 624, in main
    self.run_cmd(cfg, options)
  File "/example/.venv/lib/python3.11/site-packages/alembic/config.py", line 601, in run_cmd
    fn(
  File "/example/.venv/lib/python3.11/site-packages/alembic/command.py", line 233, in revision
    script_directory.run_env()
  File "/example/.venv/lib/python3.11/site-packages/alembic/script/base.py", line 578, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/example/.venv/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 93, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/example/.venv/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 109, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/example/src/app/migrations/env.py", line 7, in <module>
    from app.models import Base
  File "/example/src/app/models.py", line 12, in <module>
    from .models.vertex import NamedVertex, Vertex
  File "/example/src/app/models/vertex.py", line 40, in <module>
    class NamedVertex(BaseVertex):
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_api.py", line 625, in __init_subclass__
    super().__init_subclass__()
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_api.py", line 847, in __init_subclass__
    _as_declarative(cls._sa_registry, cls, cls.__dict__)
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_base.py", line 247, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_base.py", line 328, in setup_mapping
    return _ClassScanMapperConfig(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_base.py", line 563, in __init__
    self._scan_attributes()
  File "/example/.venv/lib/python3.11/site-packages/sqlalchemy/orm/decl_base.py", line 884, in _scan_attributes
    raise exc.InvalidRequestError(
**sqlalchemy.exc.InvalidRequestError: Mapper properties (i.e. deferred,column_property(), relationship(), etc.) must be declared as @declared_attr callables on declarative mixin classes.  For dataclass field() objects, use a lambda:**
```

**Versions.**
 - OS: macOS Ventura
 - Python: 3.11.4
 - Alembic: 1.12.0
 - SQLAlchemy: 2.0.21
 - Database: PostgreSQL 15
 - DBAPI: psycopg2-binary@2.9.7

**Additional context**
Not sure if this is a bug or a miss from my side. If not, any reference to related documentation will be helpful.

**Have a nice day!**

**Comment by CaselIT** [comment]
I've moved to sqlalchemy since alembic is not involved in the error

**Comment by zzzeek** [comment, marked as answer]
the error message is an informative, checked error generated by SQLAlchemy that is advising on the correct way to format the program:

     Mapper properties (i.e. deferred,column_property(), relationship(), etc.) must be declared as @declared_attr callables on declarative mixin classes.  For dataclass field() objects, use a lambda:


By "declared_attr callables" it's referring to https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.declared_attr , more generally introduced at https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html .   Specifically for special mapped constructs like `composite()` it refers to https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html#mixing-in-orm-column-property-and-other-orm-mapperproperty-classes

so here we can update your example as:

```py
import dataclasses
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, composite, mapped_column, declared_attr

class Base(MappedAsDataclass, DeclarativeBase):
    pass

@dataclasses.dataclass
class Point:
    x: int
    y: int


class BaseVertex(Base):
    __abstract__ = True

    id: Mapped[int] = mapped_column(primary_key=True, init=False)

    @declared_attr
    def start(self) -> Mapped[Point]:
        return composite(mapped_column("x1"), mapped_column("y1"))

    @declared_attr
    def end(self) -> Mapped[Point]:
        return composite(mapped_column("x2"), mapped_column("y2"))

class NamedVertex(BaseVertex):
    __tablename__ = "named_vertex"
    name: Mapped[str] = mapped_column()

```

the Table (which is what Alembic actually cares about, it knows nothing about ORM mapped classes) now has these columns:

```py
>>> from sqlalchemy.schema import CreateTable
>>> print(CreateTable(NamedVertex.__table__))

CREATE TABLE named_vertex (
	name VARCHAR NOT NULL, 
	id INTEGER NOT NULL, 
	x1 INTEGER NOT NULL, 
	y1 INTEGER NOT NULL, 
	x2 INTEGER NOT NULL, 
	y2 INTEGER NOT NULL, 
	PRIMARY KEY (id)
)

```

there's one final issue here is that you are using dataclasses, so while this constructor works:

```py
>>> nn = NamedVertex(name='name', start=Point(5, 9), end=Point(10, 12))
>>> nn
NamedVertex(id=None, name='name', start=Point(x=5, y=9), end=Point(x=10, y=12))

```

type checkers won't see this in your `__init__` method, because we are using [pep-681](https://peps.python.org/pep-0681/) to determine how this method looks:

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/8303c066-9d77-4055-b6f3-dbf8cdc53e24)


so to work around, we can use a TYPE_CHECKING block:

```py
from typing import TYPE_CHECKING


class BaseVertex(Base):
    __abstract__ = True

    id: Mapped[int] = mapped_column(primary_key=True, init=False)


    if TYPE_CHECKING:
        start: Mapped[Point]
        end: Mapped[Point]
    else:
        @declared_attr
        def start(self) -> Mapped[Point]:
            return composite(mapped_column("x1"), mapped_column("y1"))

        @declared_attr
        def end(self) -> Mapped[Point]:
            return composite(mapped_column("x2"), mapped_column("y2"))

class NamedVertex(BaseVertex):
    __tablename__ = "named_vertex"
    name: Mapped[str] = mapped_column()

```

now type checkers using pep-681 can see the constructor args (if this is important to you):

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/428f9a71-9aa7-4d73-9e5b-f4054dfc532f)


we don't yet have a syntax to make that smoother for the case of type-checked dataclasses with composites with mixins right now.

**Reply by anil-enable** [reply]
Thanks for the detailed explanation and references. It really helps. I tried it out and it works perfectly.
TYPE_CHECKING part was the icing on the cake, thanks!

**Comment by anil-enable** [comment]
I was playing around a bit with declared_attr, one observation I wanted to share w.r.t Optional

I added another composite column type `mid` to `BaseVertex` but made it Optional

```py
class BaseVertex(Base):
    ...
    @declared_attr
    def mid(self) -> Mapped[Optional[Point]]:
        return composite(mapped_column("x3"), mapped_column("y3"), default=None)
```

This leads to the NamedVertex table definition with NULL column types for x3 and y3, same is reflected in the generated migration script

```sql
CREATE TABLE named_vertex (
        name VARCHAR(255) NOT NULL, 
        id INTEGER NOT NULL, 
        x1 INTEGER NOT NULL, 
        y1 INTEGER NOT NULL, 
        x2 INTEGER NOT NULL, 
        y2 INTEGER NOT NULL, 
        x3 NULL, 
        y3 NULL, 
        PRIMARY KEY (id)
)
```

I worked around this by adding column type Integer

```py
class BaseVertex(Base):
    ...
    @declared_attr
    def mid(self) -> Mapped[Optional[Point]]:
        return composite(mapped_column("x3", Integer), mapped_column("y3", Integer), default=None)
```

This leads to the NamedVertex table definition with correct column types for x3 and y3

```sql
CREATE TABLE named_vertex (
        name VARCHAR(255) NOT NULL, 
        id INTEGER NOT NULL, 
        x1 INTEGER NOT NULL, 
        y1 INTEGER NOT NULL, 
        x2 INTEGER NOT NULL, 
        y2 INTEGER NOT NULL, 
        x3 INTEGER, 
        y3 INTEGER, 
        PRIMARY KEY (id)
)
```

**Reply by CaselIT** [reply]
this seems like a bug. do you think it's worth fixing @zzzeek ?

**Reply by zzzeek** [reply]
doesnt seem like a bug?  if you declare `Mapped[Optional[]]` then the mapped_column() will assume NULL.   that's documented for mapped_column() in general.   Adding the Integer part here, where the NULL no longer transfers over, seems more problematic to me.

as [documented](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapped-column-derives-the-datatype-and-nullability-from-the-mapped-annotation), if you want Optional but the column types are NOT NULL then set nullable=False on them.

**Reply by anil-enable** [reply]
Yes I agree, nullability is derived from Optional, issue is with data type which is also NULL instead of INTEGER (maybe i have not worded it correctly in the above example)

```sql
CREATE TABLE named_vertex (
        name VARCHAR(255) NOT NULL, 
        id INTEGER NOT NULL, 
        x1 INTEGER NOT NULL, 
        y1 INTEGER NOT NULL, 
        x2 INTEGER NOT NULL, 
        y2 INTEGER NOT NULL, 
        x3 NULL,  <-- this one
        y3 NULL,  <-- this one
        PRIMARY KEY (id)
```

**Reply by zzzeek** [reply]
oh that.  yes, that's a major bug.   Open an issue ?

**Reply by anil-enable** [reply]
ok, i will do that

**Reply by anil-enable** [reply]
Here you go #10380

Thanks
